src/Adapters/PageCategoriesRetriever.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace ProfessionalWiki\PageApprovals\Adapters;
+
+use MediaWiki\Page\PageIdentity;
+use MediaWiki\Page\WikiPageFactory;
+
+class PageCategoriesRetriever {
+
+	public function __construct(
+		private WikiPageFactory $pageFactory
+	) {
+	}
+
+	/**
+	 * @return string[]
+	 */
+	public function getPageCategories( PageIdentity $page ): array {
```

Comment:
I originally created `ApprovalAuthorizer::canApprove()` with a `PageIdentity` argument, with the idea that the API (or other classes) would check if a provided `$pageId` refers to a valid page first and handle that explicitly. The implication being after that point we can use a `PageIdentity` (i.e. a real page) instead of the `$pageId` int.

However, in `PageContentRetriever` I didn't think about that, which lead to handling the invalid/missing page (and [this comment](https://github.com/ProfessionalWiki/PageApprovals/pull/32#discussion_r1656901299)):
https://github.com/ProfessionalWiki/PageApprovals/blob/bd790f4335c685a216b742d7471025e84e98b82a/src/Adapters/PageContentRetriever.php#L17-L22

I'm not sure if we should rather have all these services take just the page ID and always handle the missing/invalid page case, or if all these services should rather take PageIdentity, which lets us skip that case.

Reply:
Using `PageIdentity` seems reasonable

Reply:
I ended up using WikiPage instead of PageIdentity. Only the former has ParserOutput, whereas the interface can be used by a Title which isn't a full page.